odoo/modules/packages.py

```diff
@@ -0,0 +1,281 @@
+    """
+    def __init__(self, name: str, package_graph: PackageGraph) -> None:
+        # manifest data
+        self.name: str = name
+        self.manifest: dict = get_manifest(name) or {}
+
+        # ir_module_module data                     # column_name
+        self.id: int = 0                            # id
+        self.state: str = 'uninstalled'             # state
+        self.dbdemo: bool = False                   # demo
+        self.installed_version: str | None = None   # latest_version (attention: Incorrect field names !! in ir_module.py)
+
+        # info for upgrade
+        self.load_state: str = 'uninstalled'        # the state when added to package_graph
+        self.load_version: str | None = None        # the version when added to package_graph
+
+        # dependency
+        self.depends: OrderedSet[Package] = OrderedSet()
+        self.package_graph: PackageGraph = package_graph
+
+        # properties which can be inherited 
+        self._properties: dict = {}
+
+    @lazy_property
+    def depth(self) -> int:
+        """ Return the longest distance from self to module 'base' along dependencies. """
+        return max(package.depth for package in self.depends) + 1 if self.depends else 0
+
+    @lazy_property
+    def phase(self) -> int:
+        if self.name == 'base':
+            return 0
+
+        if self.package_graph.mode == 'load':
+            return 1
+
+        def not_in_the_same_phase(package: Package, dependency: Package) -> bool:
+            return (package.state == 'to install') ^ (dependency.state == 'to install') or dependency.name == 'base'
+
+        return max(
+            dependency.phase + 1 if not_in_the_same_phase(self, dependency) else dependency.phase
+            for dependency in self.depends
+        )
+    
+    @lazy_property
+    def inherited_properties(self) -> dict:
+        properties = {}
+        for dependency in sorted(self.depends, key=lambda p: (p.phase, p.depth, p.name)):
+            properties.update(dependency.inherited_properties)
+        properties.update(self._properties)
+        return properties
+
+    @property
+    def demo_installable(self) -> bool:
+        return all(p.dbdemo for p in self.depends)
+
+
+class PackageGraph:
+    """
+    Sorted Python packages for Odoo modules with the same names ordered by (package.phase, package.depth, package.name)
+    """
+
+    def __init__(self, cr: BaseCursor, mode: Literal['load', 'update'] = 'load') -> None:
+        self.mode: Literal['load', 'update'] = mode
+        self._packages: dict[str, Package] = {}
+        self._cr: BaseCursor = cr
+
+    def __contains__(self, name: str) -> bool:
+        return name in self._packages
+
+    def __getitem__(self, name: str) -> Package:
+        return self._packages[name]
+
+    def __iter__(self) -> Iterator[Package]:
+        return iter(sorted(self._packages.values(), key=lambda p: (p.phase, p.depth, p.name)))
```

Comment:
Major nitpicking, this is friday evening, I wanted to have a bit of fun. Please ignore :joy: :nerd_face: 

Here calling `__iter__` is O(nlogn) and then every `__next__` is O(1). But if we go and use a heap instead, we can achieve `__iter__` in O(n) and then `__next__` in O(logn). Meaning that if we are gonna only iter on the first few items it gonna be faster (by a useless factor of O(logn)) :nerd_face: :nerd_face: :nerd_face: 

my patch doesn't work but this is friday evening and I had fun experimenting :nerd_face: :joy: 
[patch.txt](https://github.com/user-attachments/files/18540099/patch.txt)